### DISCUSSION TITLE
gRPC support for ABCI

### DISCUSSION THREAD

**Discussion by erikgrinaker** [tendermint/tendermint#5855] [General]
The gRPC ABCI client is broken due to fundamental concurrency issues in the async request/response handling in the ABCI client, i.e. `abcicli.ReqRes` -- see https://github.com/tendermint/tendermint/issues/5439#issuecomment-732098394. We should either remove the gRPC client entirely, or rewrite the `ReqRes` API.

The `socket` client does not appear to be vulnerable to this due to implementation details (i.e. it's not actually async).

Personally I think we should use gRPC as the primary ABCI protocol transport, since it has better cross-language support.

**Comment by JayT106** [comment]
https://github.com/tendermint/tendermint/issues/3483#issuecomment-852749199
